### DISCUSSION TITLE
Why is a reverse map needed for expression translation?

### DISCUSSION THREAD

**Discussion by kwaazaar** [AutoMapper/AutoMapper.Extensions.ExpressionMapping#174] [General]
When I want to map an expression on TSource to TTarget, it makes sense that my mapping configuration defines a map from TSource to TTarget. However, it also needs a map from TTarget to TSource. Why is that?

Also, mapping a certain expression, that worked on AM v6, now gives me this error, but I have no idea how to fix this: `No generic method 'Count' on type 'System.Linq.Enumerable' is compatible with the supplied type arguments and arguments. No type arguments should be provided if the method is non-generic.`

**Comment by palonza** [comment]


**Reply by kwaazaar** [reply]
Hi @palonza . Thanks for the quick reply.

It took me all day, but I managed to extract a repro project that demonstrates the issue.
The problem seems to be related to reusing a source type for a second target type. I would think this should be supported by AutoMapper, as mapping the objects directly works fine. Mapping an expression on the source is what's not working.

Hopefully my repro will be clear. You can find the code here: [https://github.com/kwaazaar/automapper-expressionmappingissue](https://github.com/kwaazaar/automapper-expressionmappingissue)

**Comment by palonza** [comment]


**Reply by kwaazaar** [reply]
Can you explain why it now works? Because I don't understand what you are doing there:
1. What is the difference between this: `cfg.CreateMap<SourceType, TargetType>().ReverseMap();` and your code:
```
cfg.CreateMap<SourceType, TargetType>();
cfg.CreateMap<TargetType, SourceType>().ReverseMap();
```
2. Why map identical types: `cfg.CreateMap<TargetChildType, TargetChildType>().ReverseMap();`
3. The last 2 lines are identical to the two above

**Comment by palonza** [comment]


**Reply by kwaazaar** [reply]
In solution 2 you change the expression to work on target types instead of source types. But my goal is to get expression translation to work, not to avoid it. Especially since this worked in AutoMapper 6.

In solution 3 you suggest to change the used types. Same here: I don't want to change the types (this repro is much simpler that the actual sources where I encounter these issues), as it's avoiding the issue, instead of solving it. Also your screenshot is about polymorphism, which does not apply here, as there is no inheritance between types (they just look-a-like).

**Comment by BlaiseD** [comment]
For most cases you only need the [TTarget to TSource](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/fd47b8413a37c76e9bc35bc28f920ca1596421ed/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs#L670) configuration when mapping an expression from TSource to TTarget.

You may also need the TSource to TTarget configuration e.g. in cases where the [expression contains a constant](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/fd47b8413a37c76e9bc35bc28f920ca1596421ed/tests/AutoMapper.Extensions.ExpressionMapping.UnitTests/XpressionMapper.Structs.Tests.cs#L109).

Issue #175  will track the failing issue (multiple destination types in expression mapping).
